ndonnx/_typed_array/funcs.py

```diff
@@ -0,0 +1,313 @@
+# Copyright (c) QuantCo 2023-2025
+# SPDX-License-Identifier: BSD-3-Clause
+from __future__ import annotations
+
+from collections.abc import Sequence
+from functools import reduce
+from types import NotImplementedType
+from typing import TYPE_CHECKING, Literal, TypeVar, overload
+
+import numpy as np
+from spox import Var
+
+from .._dtypes import DType
+from .._types import NestedSequence, PyScalar
+from .typed_array import TyArrayBase
+from .utils import promote
+
+if TYPE_CHECKING:
+    from .. import Array
+    from .._dtypes import TY_ARRAY_BASE
+    from . import onnx
+
+
+def _infer_sequence(
+    val: NestedSequence,
+) -> DType:
+    from . import onnx
```

Comment:
We should factor out whatever is causing this cyclic dependency and depend on it from`onnx` submodule (still think that's a confusing name) and from this one.

Reply:
I straightened out the dependencies. It isn't perfect now, but much better, IMHO.